compositor_pipeline/src/queue/audio_queue.rs

```diff
@@ -0,0 +1,163 @@
+        let input_queue = InputQueue::new(channels, sample_rate);
+        self.input_queues.insert(input_id, input_queue);
+    }
+
+    pub fn remove_input(&mut self, input_id: &InputId) {
+        self.input_queues.remove(input_id);
+    }
+
+    pub fn enqueue_samples(
+        &mut self,
+        input_id: InputId,
+        samples_batch: AudioSamplesBatch,
+    ) -> Result<(), QueueError> {
+        let Some(input_queue) = self.input_queues.get_mut(&input_id) else {
+            return Err(QueueError::UnknownInputId(input_id));
+        };
+        let offset = *self
+            .timestamp_offsets
+            .entry(input_id)
+            .or_insert_with(|| samples_batch.pts);
+
+        input_queue.enqueue_samples(samples_batch.samples, samples_batch.pts - offset)?;
+
+        Ok(())
+    }
+
+    pub fn pop_samples_set(&mut self, pts: Duration, length: Duration) -> AudioSamplesSet {
+        let mut samples = HashMap::new();
+        self.input_queues
+            .iter_mut()
+            .for_each(|(input_id, input_queue)| {
+                samples.insert(input_id.clone(), input_queue.pop(length));
+            });
+
+        AudioSamplesSet {
+            samples,
+            pts,
+            length,
+        }
+    }
+}
+
+enum SamplesQueue {
+    Mono(VecDeque<i16>),
+    Stereo(VecDeque<(i16, i16)>),
+}
+
+impl SamplesQueue {
+    pub fn new(channels: AudioChannels) -> Self {
+        match channels {
+            AudioChannels::Mono => SamplesQueue::Mono(VecDeque::new()),
+            AudioChannels::Stereo => SamplesQueue::Stereo(VecDeque::new()),
+        }
+    }
+
+    pub fn len(&self) -> usize {
+        match self {
+            SamplesQueue::Mono(samples) => samples.len(),
+            SamplesQueue::Stereo(samples) => samples.len(),
+        }
+    }
+}
+
+struct InputQueue {
+    queue: SamplesQueue,
+    popped_samples: u64,
+    sample_rate: u32,
+}
+
+impl InputQueue {
+    pub fn new(channels: AudioChannels, sample_rate: u32) -> Self {
+        InputQueue {
+            queue: SamplesQueue::new(channels),
+            popped_samples: 0,
+            sample_rate,
+        }
+    }
+
+    pub fn enqueue_samples(
+        &mut self,
+        samples: AudioSamples,
+        pts: Duration,
+    ) -> Result<(), QueueError> {
+        let expected_previous_samples = (pts.as_secs_f64() * self.sample_rate as f64) as i64;
+        let missing_samples =
+            expected_previous_samples - self.popped_samples as i64 - self.queue.len() as i64;
+
+        match &mut self.queue {
+            SamplesQueue::Mono(queue) => {
+                let AudioSamples::Mono(samples) = samples else {
+                    return Err(QueueError::MismatchedSamplesChannels {
+                        expected: AudioChannels::Mono,
+                        received: samples,
+                    });
+                };
+                // To account for calculation precision errors
+                if missing_samples > 5 {
```

Comment:
why 5, I wouldn't expect numeric errors to cause more than one sample duration error

Reply:
Yep, while testing I only managed to get the numeric error 1 off. I used 5 for safety. In reality, the only reason for missing samples is that the compositor couldn't receive/read some chunks of samples (e.g. lost packet). The usual length of an audio chunk in a single RTP packet is 20ms, meaning that there are hundreds of samples. I guess that's the case for all inputs (receiving or reading only a few samples is very unlikely). I can't find a probable reason for missing a few samples, but calculation errors >1 can happen. Therefore, I want to use some value >1 here. I think that anything in the range [3, 10] is a reasonable threshold here.

Reply:
With the new audio queue approach, this logic is no longer useful